### PR TITLE
Add "Search Decks" to Deck Picker

### LINKED ISSUE #7116: Deck Picker: Add "Search" Action Item
Many medical decks have a lot of sub-decks, being able to filter the list should help navigation

### PR DESCRIPTION
**Pull request by david-allison-1**
## Purpose / Description
Some users wanted this

## Fixes
Fixes #7116

## Approach

* `SearchView` and `Filter`
* If a deck matches the search string, then all children are visible
* If a deck matches the search string, then all parents are visible

## How Has This Been Tested?

On my Android 9 with the AnKing deck. Worked as expected. 
* "Create Deck" result is only visible if searched
* Decks which are not visible can be reviewed if the parent is selected.
* Nesting works as expected
* Tested on my Android Emulator
* Keyboard shortcuts no longer get in the way

image
image
image

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by mikehardy** [comment]
This is a serious feature! Nice

**Comment by david-allison-1** [comment]
Hmm... There's a slight UI redraw on the following as filtering is async

https://github.com/ankidroid/Anki-Android/blob/86c26f1d4014284c757ade154f3f930c3f5e102a/AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.java#L176-L177

I'm going to remove the notify line (as the filter will do that when complete).

**Comment by david-allison-1** [comment]
Ok, done

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 1]
```diff
@@ -2384,11 +2411,15 @@ public void __renderPage() {
...
             }
+            if (mToolbarSearchView != null) {
+                mDeckListAdapter.getFilter().filter(mToolbarSearchView.getQuery());
+            }
             // We're done here
             return;
         }
 
-        mDeckListAdapter.buildDeckList(mDueTree, getCol());
+        CharSequence filter = mToolbarSearchView != null ? mToolbarSearchView.getQuery() : null;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:2421]: Extract this above

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.java [hunk 2]
```diff
@@ -346,7 +361,87 @@ public Integer getDue() {
...
+    @Override
+    public Filter getFilter() {
+        return new DeckFilter();
+    }
+
+
+    private class DeckFilter extends Filter {
+        private ArrayList<AbstractDeckTreeNode> mFilteredDecks;
+        private DeckFilter() {
+            super();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.java:378]: Only necessary as this is a private constructor - needs no code

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.java [hunk 3]
```diff
@@ -346,7 +361,87 @@ public Integer getDue() {
...
+                    ret.add(node);
+                }
+            }
+            return ret;
+        }
+
+        @Nullable
+        private <T extends AbstractDeckTreeNode<T>> T filterDeckInternal(String filterPattern, T root) {
+
+            // If a deck contains the string, then all ts children are valid
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.java:424]:
```suggestion
            // If a deck contains the string, then all its children are valid
```

**Review by mikehardy — APPROVED** [review]
Looks good - combo of the two waits as I'll merge but you left yourself completion notes

**Pull request merged**